### DISCUSSION TITLE
Minor improvements to the Invoices screen (filtering and summation)

### DISCUSSION THREAD

**Discussion by BenGWeeks** [btcpayserver/btcpayserver#3637] [Ideas / Feature Requests]
On the Invoices screen, it would be useful to be able to easily report on the takings of the shop (without having to export to Excel). This screen could therefore be improved by:

1. Adding a "Amount (BTC)" column to the listing
2. Being able to filter by a date range (add filter to the column heading that amends the search)
3. Being able to filter by on-chain/Lightning (add filter to the column heading that amends the search)
5. Being able to filter by Status (add filter to the column heading that amends the search)
6. Adding totals to the "Amount" and "Amount (BTC)" (that sums the filtered list)

Use case: My sons runs his shop ([www.robotechy.com](www.robotechy.com)) and at the end of each month I want to total up the amount of sales he has made (in BTC) and make a transfer to his wallet on his phone (currently I have to export to Excel and number crunch a little).

**Comment by dstrukt** [comment]
Great feedback, we're going to be addressing a few of these features in the next major release, so we'll keep these in mind!

Thanks!

**Comment by ishristov** [comment]
I was just about to start the same discussion, but I'm gonna +1 this instead. Someone I'm working with is running a few gas stations and is now accepting BTC via BTCPay. But they want to see the total amount (in FIAT) that is received for the day, month and year so if 2) and 5) are implemented, that'd be very useful. @dstrukt, can you share some inside info on what is being working on? :) 🙏

**Reply by dstrukt** [reply]
Will include these comments as part of our design call, as well as share some explorative mocks + a write up addressing these points this weekend!

**Reply by pavlenex** [reply]
Basically we can showcase a toggle between store default currency and balance on the dashboard. When user clicks on the balance it switches from btc to store default currency. Will open an issue.

**Reply by pavlenex** [reply]
https://github.com/btcpayserver/btcpayserver/issues/3661

**Comment by pavlenex** [comment]
1. Not possible. Invoice is generated in the currency that you've generated it in.
2. This is already possible.
![Screenshot 2022-04-21 at 12 56 28](https://user-images.githubusercontent.com/36959754/164443803-e00837a6-effa-431b-86f4-9a16e8536dbd.png)
3. Maybe we can add a requirement to filter an invoice by payment method
4. Already possible to filter invocie by status
5. So show a balance by filtered invoices, I am not sure if this is possible invoices are created in different currencies.

So from all above.
3. Add ability to filter invoice by payment method (probably possible)
5. Add the ability to display balance for filtered invoices (don't really see how it can be done, perhaps it can only show btc balance, not local currency balances, but that then is already possible in dashboard as said in #3661

**Reply by ishristov** [reply]
5. It can be done. Get all filtered invoices -> group them by invoice currency -> show total received amount for each invoice currency.

So if we have invoices in USD only, the summary will be: "Total: 100 USD". If we have invoices in EUR and USD, the summary will be "Total: 200 EUR and 100 USD".

Also the current filter is not very straightforward to use and the UX can be improved tremendously (the post itself shows that there is a problem with the UX), e.g.:
- instead of typing storeId:... -> show dropdown with the stores' names
- instead of typing startDate, endDate -> show 2 datepickers for start and end date;
- instead of typing status -> show a multiselect dropdown with the different statuses that can be selected;
- instead of writing orderId:id or itemcode:id -> enter the id only and let the backend return all invoices that contain such id.

**Reply by BenGWeeks** [reply]
Thank you for your response @pavlenex :-)

> 1. Not possible. Invoice is generated in the currency that you've generated it in.

I'm not sure why I follow this isn't possible, as you convert the fiat total to BTC at the point of sale? In fact, when I come to transfer the amount to my son at the end of the month, the GBP total (by exporting and summing this column) is no longer applicable as the amount is stored in BTC on the node and the exchange rate is different since then, so I need to know the BTC amount to transfer to him and not fiat. Actually, if I click on "Details" it's the value from the "Paid" amount (whether on-chain or off-chain) that would be great as it's own column.

![image](https://user-images.githubusercontent.com/5950171/164460686-05ad2ebd-259d-47af-aab4-dea5fa76e827.png)

> 2. This is already possible.

It is, but it's not a user-friendly way to filter (in my opinion). As a user, my eye jumped over the search input, and I was looking for filtering at the top of the columns (see crude depiction below).

![image](https://user-images.githubusercontent.com/5950171/164459226-9f7c62de-3d69-43af-a8cf-2fae7f11fb81.png)

**Reply by pavlenex** [reply]
I don't agree with what's proposed on multicurrency, but suggestion for invoice filtering seem to be spot on, hopefully someone can tackle those in the future.

**Reply by BenGWeeks** [reply]
> I don't agree with what's proposed on multicurrency, but suggestion for invoice filtering seem to be spot on, hopefully someone can tackle those in the future.

Why's that @pavlenex? I'm intrigued :-) In my view the fiat currency and BTC value are equally valuable. The fiat currency amount was the invoice amount at the time of the sale, but that doesn't reflect the the current value of the sale, that is the BTC amount as the exchange rate has changed since the invoice and it is BTC that we receive.

**Reply by pavlenex** [reply]
Because it's a mess to handle it the way it was proposed. If you care about fiat value of your holdings, those should be reflected via dashobard. Dynamically shifting values in the invoice list can lead to confusion.

**Reply by BenGWeeks** [reply]
> Because it's a mess to handle it the way it was proposed. If you care about fiat value of your holdings, those should be reflected via dashobard. Dynamically shifting values in the invoice list can lead to confusion. 

Actually, I don't care about GBP but I do care about BTC holding (but others may differ) :-) Although for reconciliation with orders, it is useful to know original amount. Do you think the user should be able to switch between invoice fiat amount, and current BTC amount perhaps instead?

**Reply by BenGWeeks** [reply]
And the dashboard doesn't reflect my fiat holdings? In the example above, we invoiced £61.50, but that is not the current value of my holding - that was the value at the point of invoicing. Actually that 0.00194106 BTC is worth approximately £63.64 if I was to exchange it today.

**Reply by pavlenex** [reply]
There's a reason invoices are fixed value. Invoice page is not a place that reflects your balance. You receive payments to your wallet through invoices, but a payment can be received without an invoice.

Dashboard can reflect your payment method balance, in your case on and off chain.

**Reply by BenGWeeks** [reply]
Ah, I get you. Yes, I suppose this page is titled "Invoices" and I am using it as a way to see "Transactions" it could be argued.

**Reply by ishristov** [reply]
@pavlenex, keep in mind that if we want this (btcpay) to be used by real businesses (and not crypto punks or degenerates only), those businesses would want to pay taxes on the btc they receive. To do that, they should have a super easy way in the UI to see what was the value they received in FIAT terms, because that's how taxes work in every jurisdiction that I know of. You receive 0.01 BTC but you pay taxes on the FIAT value on that BTC on that particular day. If we tell them to export CSVs and import them into excel to see the total fiat amount received per day/month/year, that simply won't happen. Honestly, I can't comprehend why would you oppose that.

You say that the user can see the "value of their holdings in fiat" in the dashboard, but that's not the same. Taxes are owed on the fiat value on the received payments (only the invoices have that info atm), not on the balance.
